ttg/ttg/parsec/ttg.h

```diff
@@ -1494,87 +1494,77 @@ namespace ttg_parsec {
-        parsec_gpu_task_t *gpu_task;
-        /* PaRSEC wants to free the gpu_task, because F***K ownerships */
-        gpu_task = static_cast<parsec_gpu_task_t*>(std::calloc(1, sizeof(*gpu_task)));
-        PARSEC_OBJ_CONSTRUCT(gpu_task, parsec_list_item_t);
-        gpu_task->ec = parsec_task;
-        gpu_task->task_type = 0; // user task
-        gpu_task->last_data_check_epoch = 0; // used internally
-        gpu_task->pushout = 0;
-        gpu_task->submit = &TT::device_static_submit<Space>;
-
-        // one way to force the task device
-        // currently this will probably break all of PaRSEC if this hint
-        // does not match where the data is located, not really useful for us
-        // instead we set a hint on the data if there is no hint set yet
-        //parsec_task->selected_device = ...;
-
-        /* set the gpu_task so it's available in register_device_memory */
-        task->dev_ptr->gpu_task = gpu_task;
-
-        /* TODO: is this the right place to set the mask? */
-        task->parsec_task.chore_mask = PARSEC_DEV_ALL;
 
-        /* copy over the task class, because that's what we need */
-        task->dev_ptr->task_class = *task->parsec_task.task_class;
-
-        // first invocation of the coroutine to get the coroutine handle
-        static_op<Space>(parsec_task);
-
-        /* when we come back here, the flows in gpu_task are set (see register_device_memory) */
-
-        parsec_task_class_t& tc = task->dev_ptr->task_class;
-
-        // input flows are set up during register_device_memory as part of the first invocation above
-        for (int i = 0; i < MAX_PARAM_COUNT; ++i) {
-          tc.in[i]  = gpu_task->flow[i];
-          tc.out[i] = gpu_task->flow[i];
+      /* set up the device task */
+      parsec_gpu_task_t *gpu_task = task->dev_ptr->gpu_task;
+      /* TODO: needed? */
+      std::memset(gpu_task, 0, sizeof(gpu_task));
+      /* construct the GPU task */
+      PARSEC_OBJ_CONSTRUCT(gpu_task, parsec_list_item_t);
+      gpu_task->ec = parsec_task;
+      gpu_task->task_type = 0; // user task
+      gpu_task->last_data_check_epoch = 0; // used internally
+      gpu_task->pushout = 0;
+      gpu_task->submit = &TT::device_static_submit<Space>;
+      gpu_task->release_device_task = &release_device_task;
+
+      /* TODO: is this the right place to set the mask? */
+      task->parsec_task.chore_mask = PARSEC_DEV_ALL;
+
+      /* copy over the task class, because that's what we need */
+      task->dev_ptr->task_class = *task->parsec_task.task_class;
+
+      // first invocation of the coroutine to get the coroutine handle
+      static_op<Space>(parsec_task);
+
+      /* when we come back here, the flows in gpu_task are set (see register_device_memory) */
+
+      parsec_task_class_t& tc = task->dev_ptr->task_class;
+
+      // input flows are set up during register_device_memory as part of the first invocation above
+      for (int i = 0; i < MAX_PARAM_COUNT; ++i) {
```

Comment:
Why is the upper bound here always `MAX_PARAM_COUNT` ?

Reply:
Because we don't know how many device inputs the application will give us. We could put a stop there but the impact will be marginal.